### PR TITLE
[ET-1186] Enhancement - Add loader/spinner on Ticket Commerce Checkout page

### PR DESCRIPTION
**Pull request by rafsuntaskin**
### 🎫 Ticket

[ET-1186] 

### 🗒️ Description

* Added a loader component on the Ticket Commerce JS.
* Created custom events that can be fired from the Gateway scripts to show/hide the loader.
* Added loader triggers to PayPal checkout JS to show and hide loader view.
* Load PayPal script in async.
* Add Ticket Commerce JS as a dependency for PayPal checkout JS.

### 🎥 Artifacts 
🎥 screencast: 
 * ~~Take 1 : https://d.pr/v/92H5qx~~
 * Take 2 : https://d.pr/v/vQPThq ( latest with JS trigger )

### ✔️ Checklist
- [ ] ~~I've included a changelog entry both in readme.txt and changelog.txt files.~~ Not required 
- [x] My code is tested. 
- [x] My code has [proper inline documentation](https://the-events-calendar.github.io/products-engineering/docs/code-standards/).

[ET-1186]: https://theeventscalendar.atlassian.net/browse/ET-1186

### CODE REVIEW

File: src/resources/js/v2/tickets-commerce.js
Comment (juanfra) [inline src/resources/js/v2/tickets-commerce.js:186]: We already have the implementation of `tickets-loader.js` that we can make use of. Here's the library: https://github.com/the-events-calendar/event-tickets/blob/master/src/resources/js/v2/tickets-loader.js

And here's an example of how this can be used from other places: https://github.com/the-events-calendar/event-tickets/blob/master/src/resources/js/v2/rsvp-manager.js#L178

We need to be sure that the tickets commerce script uses tickets loader as a dependency and then we can make use of the library instead of re-implementing the loader.

File: src/resources/js/v2/tickets-commerce.js
Comment (bordoni) [inline src/resources/js/v2/tickets-commerce.js:40]: Let's make sure we use `tecTicketsCommerce` no need for `Checkout` there too.
For jQuery events, the namespace comes after the event name. So it would be :
* `'showLoader.tecTicketsCommerce'`

File: src/resources/js/commerce/gateway/paypal/checkout.js
Comment (bordoni) [inline src/resources/js/commerce/gateway/paypal/checkout.js:364]:
```suggestion
		$document.on( 'DOMNodeInserted', obj.selectors.buttons, function() {
			$document.trigger( tribe.tickets.commerce.customEvents.hideLoader );
		} );
```

File: src/Tickets/Commerce/Shortcodes/Checkout_Shortcode.php
Comment (bordoni) [inline src/Tickets/Commerce/Shortcodes/Checkout_Shortcode.php:56]: No need to pass it down on this case, use it directly on the template.

**Review by bordoni — CHANGES_REQUESTED** [review]

File: src/views/v2/components/loader/loader.php
Comment (juanfra) [inline src/views/v2/components/loader/loader.php:22]: Why are we modifying the default behavior of the loader? The norm is that by default it's hidden. This change can potentially cause issues in other places where we use the loader. 

I feel that if we actually need this to be visible by default in a particular scenario we can do it with the JS method directly.

File: src/views/v2/components/loader/loader.php
Comment (juanfra) [inline src/views/v2/components/loader/loader.php:26]: We could avoid the if by doing something like this:

```
$spinner_classes = [
	'tribe-common-a11y-hidden => empty( $visible ),
]
```

(as part of the definition of the `$spinner_classes` above).

File: src/views/v2/components/loader/loader.php
Comment (juanfra) [inline src/views/v2/components/loader/loader.php:18]:
```suggestion
$visible        = tribe_is_truthy( $this->get( 'visible' ) );
```
This method already returns a bool, I think there's no need to have the OR false.

File: src/resources/js/v2/tickets-commerce.js
Comment (juanfra) [inline src/resources/js/v2/tickets-commerce.js:171]: Nice use of the library! 🎉

File: src/resources/js/commerce/gateway/paypal/checkout.js
Comment (juanfra) [inline src/resources/js/commerce/gateway/paypal/checkout.js]: If we keep the loader template as it was, we can do a `show` here. That's how we do it for the other places where we make use of the loader.

**Review by juanfra — COMMENTED** [review]
Great use of the loader library! A couple of questions regarding the modification of a lib that can have effects on other places.

File: src/resources/js/commerce/gateway/paypal/checkout.js
Comment (rafsuntaskin) [reply src/resources/js/commerce/gateway/paypal/checkout.js]: Yes, that is what I thought until I found that we are not getting into the `ready` state while the `PayPal Checkout` script is getting loaded. So, it keeps showing blank and by the time we trigger the loader to show the buttons are already rendered and the loader doesn't help at all.

So, I have modified the loader to be visible at page render immediately while the JS scripts get loaded.

File: src/views/v2/components/loader/loader.php
Comment (rafsuntaskin) [reply src/views/v2/components/loader/loader.php:22]: We are not getting into the ready state while the PayPal Checkout script is getting loaded. So, it keeps showing blank and by the time we trigger the loader to show, the Paypal buttons are already rendered and the loader doesn't help at all.

So, I have modified the loader to be visible at page render immediately while the JS scripts get loaded.

The loader template should be unaffected as we are changing the classes only if the `visible` variable is set.

Comment (bordoni) [reply src/views/v2/components/loader/loader.php:22]: I would still avoid changing the default template and make the changes by overwriting this from the Usage of the component

File: src/resources/js/v2/tickets-commerce.js
Comment (paulmskim) [inline src/resources/js/v2/tickets-commerce.js:181]: let's move these above so that we define first and use them after. `bindLoaderEvents` uses `loaderShow` and `loaderHide`, so those should be defined first, and `bindCheckoutEvents` uses `bindLoaderEvents`, so we should move `bindLoaderEvents` above.

File: src/views/v2/components/loader/loader.php
Comment (paulmskim) [reply src/views/v2/components/loader/loader.php:22]: This change shouldn't affect the other uses of the loader already? Since we are not passing in a `visible` property to the other loaders, it should still print the hidden class.

However, I do think that adding the hidden class by default is not the right call here, we should allow users of this template to pass it in on their own (there is now no way to show by default). Is there a way we can change that in the places we need to? Or does that touch too many pieces?

Comment (juanfra) [reply src/views/v2/components/loader/loader.php:22]: > Is there a way we can change that in the places we need to? Or does that touch too many pieces?

At least the tickets block, RSVP, AR page, AR modal, Manual Attendees.

With regards to the default behavior (hidden), ET is following [what TEC does](https://github.com/the-events-calendar/the-events-calendar/blob/master/src/views/v2/components/loader.php#L17), so that's kind of in-line.

Comment (paulmskim) [reply src/views/v2/components/loader/loader.php:22]: Yeah, not the best implementation :(, we hadn't thought about that.

Comment (paulmskim) [reply src/views/v2/components/loader/loader.php:22]: The other thing we could do is in the JS, on ready, we can remove the `hidden` class.

**Comment by rafsuntaskin** [comment]
I have updated the scripts to show the loader using JS now. Opted out from the default HTML show.

* Load PayPal script in async.
* Add Ticket Commerce JS as a dependency for PayPal checkout JS.

Once I get approval for the new changes, then I will revert the changes in the `Loader` component that was made for showing loader immediately in HTML.

File: src/resources/js/v2/tickets-commerce.js
Comment (bordoni) [inline src/resources/js/v2/tickets-commerce.js:172]:
```suggestion
		$document.trigger( 'beforeSetup.tecTicketsCommerce', [ $container ] );
```

File: src/resources/js/v2/tickets-commerce.js
Comment (bordoni) [inline src/resources/js/v2/tickets-commerce.js:180]:
```suggestion
		$document.trigger( 'afterSetup.tecTicketsCommerce', [ $container ] );
```

**Review by bordoni — CHANGES_REQUESTED** [review]

File: src/resources/js/v2/tickets-commerce.js
Comment (bordoni) [reply src/resources/js/v2/tickets-commerce.js:180]: When we change these particular filters we modify `commerce/gateway/paypal/checkout.js` that make usage of these.

File: src/resources/js/v2/tickets-commerce.js
Comment (bordoni) [reply src/resources/js/v2/tickets-commerce.js:172]: When we change these particular filters we modify `commerce/gateway/paypal/checkout.js` that make usage of these.

File: src/resources/js/commerce/gateway/paypal/checkout.js
Comment (paulmskim) [inline src/resources/js/commerce/gateway/paypal/checkout.js:381]: curious why this needs to be done on `onload` rather than just in the `ready` function?

Comment (paulmskim) [inline src/resources/js/commerce/gateway/paypal/checkout.js:381]: Should this check whether the `checkoutContainer` exists before running any of this code? Saves us from running code that doesn't need to be run in the admin.

File: src/resources/js/commerce/gateway/paypal/checkout.js
Comment (paulmskim) [inline src/resources/js/commerce/gateway/paypal/checkout.js:366]: make the function a separate named function on the `obj` and remove the `DOMNodeInserted` event listener after triggering the `hideLoader` event.

File: src/resources/js/commerce/gateway/paypal/checkout.js
Comment (rafsuntaskin) [reply src/resources/js/commerce/gateway/paypal/checkout.js:381]: We are loading the external PayPal script using the template: https://github.com/the-events-calendar/event-tickets/blob/1a756c95dedbebcdb72bbd0b1e81302d7276382c/src/views/v2/commerce/gateway/paypal/checkout-script.php#L29-L28

So, the `ready` method triggers before PayPal is available causing an error.

Comment (rafsuntaskin) [reply src/resources/js/commerce/gateway/paypal/checkout.js:381]: the `tribe.tickets.commerce` JS is a dependency for this script. So, that gets loaded before this `ready` is fired.

We are only using this script in FE, so I am not sure where it can run in admin.

Comment (paulmskim) [reply src/resources/js/commerce/gateway/paypal/checkout.js:381]: gotcha. could we maybe target that specific element for the `onload` event instead of the entire window?

File: src/views/v2/components/loader/loader.php
Comment (juanfra) [reply src/views/v2/components/loader/loader.php:22]: Yep, that's what I suggested initially, and here's what Rafsun found: https://github.com/the-events-calendar/event-tickets/pull/2079#discussion_r705233563

**Review by bordoni — APPROVED** [review]

File: src/resources/js/v2/tickets-commerce.js
Comment (juanfra) [inline src/resources/js/v2/tickets-commerce.js:180]: Maybe as a suggestion, to add more clarity, and to make it explicit that the trigger corresponds to the generic before and after setup, I'd create a method called `bindEvents` with the two triggers, and calling `bindLoaderEvents` and `bindCheckoutEvents` and just keeping `bindCheckoutItemDescriptionToggle` in `bindCheckoutEvents`.

**Review by juanfra — APPROVED** [review]
One small suggestion, other than that it's looking great. Good job!!! ⚡

File: src/resources/js/v2/tickets-commerce.js
Comment (rafsuntaskin) [reply src/resources/js/v2/tickets-commerce.js:180]: @juanfra Sorry, I am not clear about this one. `bindLoaderEvents` is part of the checkout and that's why it's within the `bindCheckoutEvents `. Which generic triggers you are referring to?

**Review by juanfra — APPROVED** [review]

**Pull request merged**
